File: 0725-split-linked-list-in-parts/0725-split-linked-list-in-parts.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def splitListToParts(self, head: Optional[ListNode], k: int) -> List[Optional[ListNode]]:
        res = [None for _ in range(k)]
        
        n = 0
        curr = head
        while curr:
            n += 1
            curr = curr.next
            
        l, rem = divmod(n, k)
        
        curr = head
        
        for i in range(k):
            part = curr
            size = l + (1 if i < rem else 0)
            
            for _ in range(size - 1):
                if curr:
                    curr = curr.next
                    
            if curr:
                temp = curr.next
                curr.next = None
                curr = temp
                
            res[i] = part
            
        return res

File: 0725-split-linked-list-in-parts/0725-split-linked-list-in-parts.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode[] splitListToParts(ListNode head, int k) {
        ListNode[] arr = new ListNode[k];
        
        int l = 0;
        ListNode curr_t = head;
        
        while (curr_t != null){
            curr_t = curr_t.next;
            l++;
        }
        
        int d = l / k;
        int rem = l % k;
        
        ListNode curr = head;
        
        for (int i=0; i < k; i++){
            int size = d + (i < rem ? 1 : 0);
            arr[i] = curr;
            
            for (int j=0; j<size - 1; j++){
                if (curr != null) curr = curr.next;
            }
            
            if (curr != null){
                ListNode temp = curr.next;
                curr.next = null;
                curr = temp;
            }
        }
        
        return arr;
    }
}